controller/controller_server/utils.py

```diff
@@ -8,42 +10,153 @@
 import controller.controller_server.messages as messages
 from controller.array_action.config import FC_CONNECTIVITY_TYPE, ISCSI_CONNECTIVITY_TYPE
 from controller.common.csi_logger import get_stdout_logger
+from controller.controller_server.controller_types import Secret
 from controller.controller_server.errors import ObjectIdError, ValidationException
 from controller.csi_general import csi_pb2
 
 logger = get_stdout_logger()
 
 
-def get_array_connection_info_from_secret(secrets):
-    user = secrets[config.SECRET_USERNAME_PARAMETER]
-    password = secrets[config.SECRET_PASSWORD_PARAMETER]
-    array_addresses = secrets[config.SECRET_ARRAY_PARAMETER].split(",")
+def _is_topology_match(secret_topologies, node_topologies):
+    for topologies in secret_topologies:
+        logger.debug(
+            "Comparing topologies: object topologies: {},"
+            " node topologies: {}".format(topologies, node_topologies))
+        if all(item in node_topologies.items() for item in topologies.items()):
+            return True
+    return False
+
+
+def get_volume_topologies(request):
+    if request.HasField(config.REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD):
+        accessibility_requirements = request.accessibility_requirements
+        if accessibility_requirements.preferred:
+            topologies = accessibility_requirements.preferred[0].segments
+            logger.info("Chosen volume topologies: {}".format(topologies))
+            return topologies
+
+
+def get_secret_by_topologies(secret_config_string, node_topologies):
+    secret_config = ast.literal_eval(secret_config_string)
+    for secret_uid, secret_info in secret_config.items():
+        secret_topologies = secret_info.get(config.SECRET_SUPPORTED_TOPOLOGIES_PARAMETER)
+        if _is_topology_match(secret_topologies, node_topologies):
+            return secret_info, secret_uid
+
+
+def get_secret_by_uid(secret_config_string, secret_uid):
+    secret_config = ast.literal_eval(secret_config_string)
+    return secret_config.get(secret_uid)
+
+
+def get_pool_by_uid(pools, uid):
```

Comment:
consider renaming `pools` to something like `raw_pools_by_<final_name_of_secret_uid>`

Reply:
raw? can you explain?

Reply:
i.e. the raw, unprocessed, unparsed string

Reply:
no need anymore